### PR TITLE
Increase Network Policy tests coverage

### PR DESCRIPTION
**Pull request by MaysaMacedo**
**What this PR does / why we need it**:
Validates correct behavior when :

- NetworkPolicyIngressRule is updated
- Network Policy is enforced and Namespace label is updated
- Network Policy is enforced and Pod label is updated
- Network Policy is enforced based on NamespaceSelector and named port

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```
/kind feature
/area testing
/sig-testing

### CODE REVIEW

**Comment by MaysaMacedo** [comment]
@kubernetes/sig-network-pr-reviews

**Comment by vllry** [comment]
/assign @vllry

Will review when I get to the office. :)

**Comment by MaysaMacedo** [comment]
Thank you, @vllry . :)

File: test/e2e/network/network_policy.go
Comment (vllry) [inline test/e2e/network/network_policy.go]: Compile error due to this import: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/78177/pull-kubernetes-bazel-build/1134114894237405184/build-log.txt

Requires explicit dependency listing, not transitive dependencies.

File: test/e2e/network/network_policy.go
Comment (vllry) [inline test/e2e/network/network_policy.go]: The test name should be more generic - EG `should allow ingress access from namespace on one named port`.

File: test/e2e/network/network_policy.go
Comment (vllry) [inline test/e2e/network/network_policy.go]: Incorrect test message?

File: test/e2e/network/network_policy.go
Comment (vllry) [inline test/e2e/network/network_policy.go]: Why use 3 clients when you only test with 2 at each stage?

File: test/e2e/network/network_policy.go
Comment (vllry) [inline test/e2e/network/network_policy.go]: This is nitpicky, but I would prefer variables over magic numbers (EG clientAAllowedPort = 80). Makes the code easier to read without cross-referencing. Not sure if this follows or conflicts with official style though?

File: test/e2e/network/network_policy.go
Comment (vllry) [inline test/e2e/network/network_policy.go]: Same comment about test name, the name should be the functionality tested, without any implementation details ("client-b").

**Comment by vllry** [comment]
Thanks for submitting @MaysaMacedo! There's a few small issues and the build needs fixing.

File: test/e2e/network/network_policy.go
Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: When I used only two options of client names some errors occurred due to a client pod with same name as the one in creation  be in terminating status. But, I managed to reorganize to only use   two clients names.

Comment (vllry) [reply test/e2e/network/network_policy.go]: Is there still a concern around this timing condition? Flaking tests are a big issue in the Kubernetes codebase, so we try to either explicitly wait for events, or don't share resources.

Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: I've updated the implementation to explicit wait for the pod deletion event to be done.

File: test/e2e/network/network_policy.go
Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: I'm not sure if there is any official style for this case, I couldn't find any. If there is some I would really appreciate if you could point me to it.
I also think is better to have variables, I was trying to follow the current style for Network Policies tests.
I took a look on some others network tests, I could find some that do not use ports as magic numbers, such as [kubeproxy](https://github.com/MaysaMacedo/kubernetes/blob/add-more-e2e-np-tests/test/e2e/network/kube_proxy.go#L44) and  [networking](https://github.com/MaysaMacedo/kubernetes/blob/add-more-e2e-np-tests/test/e2e/network/networking.go#L109).

**Comment by MaysaMacedo** [comment]
Thank you for the review @vllry .

File: test/e2e/network/network_policy.go
Comment (vllry) [reply test/e2e/network/network_policy.go]: If you don't mind! I won't block the PR based on this, it's just a recommendation.

Unfortunately, outside of the API, there aren't very strong style guidelines that I know of.

**Comment by vllry** [comment]
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/testing.md this guide might be useful for running some tests locally - it's faster feedback than waiting for Prow.

File: test/e2e/network/network_policy.go
Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: Just updated. Thanks!

**Comment by vllry** [comment]
/test pull-kubernetes-e2e-gce
That one looks like an infra flake.

File: test/e2e/network/network_policy.go
Comment (vllry) [inline test/e2e/network/network_policy.go]: Magic number snuck in there still ;)

File: test/e2e/network/network_policy.go
Comment (vllry) [inline test/e2e/network/network_policy.go]: This can be removed, as testCannotConnect will add it.

File: test/e2e/network/network_policy.go
Comment (vllry) [inline test/e2e/network/network_policy.go]: Would it be faster to use unique names, rather than waiting for the teardown to reuse the name? (I suspect so)

Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: Yes, it would be faster, but it would not cover the corner cases generated when updating a policy rule to use a different `podSelector`. The name given to a pod is also used for labeling it. Thus, I believe is good to ensure that, for example, when the `podSelector` on a `NetworkPolicyIngressRule` is updated the pod that was previous selected is not selected anymore.

Comment (vllry) [reply test/e2e/network/network_policy.go]: I see. A bit unfortunate, but not worth blocking on.

**Comment by vllry** [comment]
/lgtm

@johnbelamaric ready for a look :)

File: test/e2e/network/network_policy.go
Comment (oomichi) [inline test/e2e/network/network_policy.go]: The above 3 lines (line 444-446) can be merged into a single `framework.ExpectNoError(err, "Expected pod to be not found.")`

File: test/e2e/network/network_policy.go
Comment (oomichi) [inline test/e2e/network/network_policy.go]: ditto: for line 450-452 into a single call.

File: test/e2e/network/network_policy.go
Comment (oomichi) [inline test/e2e/network/network_policy.go]: ditto: for line 491-493.

File: test/e2e/network/network_policy.go
Comment (oomichi) [inline test/e2e/network/network_policy.go]: The above 3 lines (line 439-441) can be merged into a single framework.ExpectNoError(err, "Expected pod to be not found.")

File: test/e2e/network/network_policy.go
Comment (oomichi) [inline test/e2e/network/network_policy.go]: nit: Why we need to separate this single API call into multiple lines? Just curious because I never see this pattern on a single API call.

Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: Just followed the pattern I've seen in another file. But, it can avoid having a big line, which is not actually the case in here, so I've updated to a single line.

**Comment by MaysaMacedo** [comment]
Thanks for the review @oomichi  .

**Review by oomichi — COMMENTED** [review]
Thanks for updating.
As https://kubernetes.slack.com/archives/CGXR5GX6Z/p1560200302009800?thread_ts=1560197588.009100&cid=CGXR5GX6Z @MaysaMacedo has confirmed this test works fine on local environment.

/lgtm

**Comment by MaysaMacedo** [comment]
@vllry @oomichi Could you take another look in this PR? I had to do a rebase. Thanks!

**Comment by oomichi** [comment]
Thanks

/lgtm

**Comment by bowei** [comment]
cc: @lzang

File: test/e2e/network/network_policy.go
Comment (aanm) [inline test/e2e/network/network_policy.go]: Seems odd the name describes that client-a should be allowed but on line 535 it is testing that client-a should not connect. I guess the policy name is the incorrect part here.

File: test/e2e/network/network_policy.go
Comment (aanm) [inline test/e2e/network/network_policy.go]: Value can be omitted for some operations. consider having `json:"value,omitempty"` instead

**Review by aanm — CHANGES_REQUESTED** [review]
2 small nits, very good addition @MaysaMacedo

File: test/e2e/network/network_policy.go
Comment (lzang) [inline test/e2e/network/network_policy.go]: Shouldn't we clean up the namespace when the test finished?

Comment (MaysaMacedo) [reply test/e2e/network/network_policy.go]: Namespaces are being deleted after test is completed.
[kubernetes/test/e2e/framework/framework.go](https://github.com/kubernetes/kubernetes/blob/a404238d1c90ee00827624249b9223e7644e3f99/test/e2e/framework/framework.go#L404)

File: test/e2e/network/network_policy.go
Comment (aanm) [reply test/e2e/network/network_policy.go]: @lzang right, so it should be `allow-client-b-via-named-port-ingress-rule`

**Comment by MaysaMacedo** [comment]
Thanks for the review @aanm @lzang .

**Comment by lzang** [comment]
Thanks Maysa! 
/lgtm

**Review by aanm — APPROVED** [review]

**Comment by cmluciano** [comment]
Need an approver 

/assign @danwinship @caseydavenport

**Pull request merged**
